flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -593,6 +593,10 @@ else if (current == ExecutionState.CANCELING) {
 			LOG.info("Loading JAR files for task {}.", this);
 
 			userCodeClassLoader = createUserCodeClassloader();
+
+			// make sure the user code classloader is accessible thread-locally
+			executingThread.setContextClassLoader(userCodeClassLoader);
```

Comment:
Maybe moving this just before
```
invokable = loadAndInstantiateInvokable(userCodeClassLoader, nameOfInvokableClass, env);
```
would be more clean and safe? And maybe add some comment like
```
// We are setting the correct context class loader for the whole interaction with the `invokable`
```
? 

The contract that we would introduce here, is that user code should be executed with the `userCodeClassLoader`. Before constructing the invokable, we are not touching any user code, so there should be no point (and could potentially do some harm) in setting context class loader so early.

Arguably this is a separate bug fix that just wasn't discovered before.

Reply:
> Arguably this is a separate bug fix that just wasn't discovered before.

I can keep the change as a separate commit that I add before the other commits.

Reply:
Imo, the issue with class loader is happening to the change that introduced in this PR.

This happens, because you are eagerly creating `LegacySourceFunctionThread` and it inherits 
default class loader. So to fix your issue, you are either need to set class loader in `LegacySourceFunctionThread.run()` or on the thread that instantiates the `invokable`.

Instead of setting the `userCodeClassLoader` eagerly, I propose to revert any changes related to it. You can instantiate the `LegacySourceFunctionThread` field reference in the `SourceStreamTask` later: 
 * as it was instantiated before in `SourceStreamTask::performDefaultAction()`;
 * or move the `LegacySourceFunctionThread` creation to `SourceStreamTask::init()`.

Reply:
**Update:** it looks like that in this context (`Task::doRun()`) `executingThread` is supposed to be the same as `Thread.currentThread()`.
What about changing `executingThread.setContextClassLoader(userCodeClassLoader)` to `Thread.currentThread().setContextClassLoader(userCodeClassLoader)` just before
```
invokable = loadAndInstantiateInvokable(...);
```
?
This also may help some tests, that call `Task.run()` directly.